Update RNN PTB example to use state_is_tuple

jihunchoi opened the pull request:
Currently, the PTB example emits warnings that using without `state_is_tuple` would soon be deprecated.
It can confuse newbies (like me), so it would be nice to use the one that makes no warning.

jihunchoi committed:
Update RNN PTB example to use state_is_tuple

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented:
ping for @ebrevdo

vrv commented:
ping for this too!  thanks

ebrevdo commented:
Jenkins, test this please.

vrv commented:
Good to merge?

ebrevdo commented:
Still some issues. Will review today.

dipanjanS commented:
Hi, any updates on this? wondering if the proposed changes are alright to use since the latest version of TensorFlow emits warning as mentioned.

ebrevdo commented on tensorflow/models/rnn/ptb/ptb_word_lm.py:
```diff
@@ -239,26 +239,39 @@ class TestConfig(object):
-                                 {m.input_data: x,
-                                  m.targets: y,
-                                  m.initial_state: state})
+  state = []
```
you should be able to use state = session.run(model.initial_state)

does that work at head?

ebrevdo commented on tensorflow/models/rnn/ptb/ptb_word_lm.py:
```diff
@@ -239,26 +239,39 @@ class TestConfig(object):
+  for step, (x, y) in enumerate(reader.ptb_iterator(data, model.batch_size,
+                                                    model.num_steps)):
+    fetches = []
+    fetches.append(model.cost)
```
i think you can do:

fetches = [model.cost, model.final_state, eval_op]

then

cost, state, _ = session.run(fetches, feed_dict)

though i wouldn't use state since it conflicts with the variable state above.

dipanjanS replied:
Hi @ebrevdo ,

Thanks for the suggestion, this works till the point where I am having some issues with the state, since we have `m.initial_state: state` in `feed_dict` but now our initial state is no longer a concatenated tensor but two different ones for the `c` and `h` states so how can we pass it to the `feed_dict`? 

I just started with TensorFlow recently so I might be missing something obvious.

ebrevdo replied:
Looks like you have to feed those separately.

On Jul 28, 2016 3:47 AM, "Dipanjan Sarkar" <notifications@github.com> wrote:

> In tensorflow/models/rnn/ptb/ptb_word_lm.py
> <https://github.com/tensorflow/tensorflow/pull/2580#discussion_r72601158>:
>
> >    start_time = time.time()
> >    costs = 0.0
> >    iters = 0
> > -  state = m.initial_state.eval()
> > -  for step, (x, y) in enumerate(reader.ptb_iterator(data, m.batch_size,
> > -                                                    m.num_steps)):
> > -    cost, state, _ = session.run([m.cost, m.final_state, eval_op],
> > -                                 {m.input_data: x,
> > -                                  m.targets: y,
> > -                                  m.initial_state: state})
> > +  state = []
> > +  for c, m in model.initial_state: # initial_state: ((c1, m1), (c2, m2))
> > +    state.append((c.eval(), m.eval()))
> > +  for step, (x, y) in enumerate(reader.ptb_iterator(data, model.batch_size,
> > +                                                    model.num_steps)):
> > +    fetches = []
> > +    fetches.append(model.cost)
>
> Hi @ebrevdo <https://github.com/ebrevdo> ,
>
> Thanks for the suggestion, this works till the point where I am having
> some issues with the state, since we have m.initial_state: state in
> feed_dict but now our initial state is no longer a concatenated tensor
> but two different ones for the c and h states so how can we pass it to
> the feed_dict?
>
> I just started with TensorFlow recently so I might be missing something
> obvious.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/2580/files/2819a680c951220d30b47b2212535412a66bbe7a#r72601158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABtimzBJOGYRRBqB4_F3hyBrqinFmwr9ks5qaIjGgaJpZM4IpzVZ>
> .
>

jihunchoi commented:
Hello, thank you for the comments! I will look through them in a few days.

dipanjanS commented:
Thanks a lot, this works fine, just used the same for modeling on sequences

zsdonghao commented:
Hi , there are a ``tutorial_ptb_lstm_state_is_tuple`` in TensorLayer repo, hope it help. 
https://github.com/zsdonghao/tensorlayer

danmane commented:
ping @jihunchoi

jihunchoi committed:
Merge branch 'master' into ptb_use_state_tuple

jihunchoi committed:
Modify RNN PTB example codes to be more clear

jihunchoi commented:
Hello, sorry for the late response.
I have been apart from the library for weeks...
According to the comments, I modified codes to be more clear.
I removed messy index-magics by using LSTMStateTuple; look line 260 and 263-270.

jihunchoi committed:
Fix typo

jihunchoi commented:
Fixed a stupid typo.

ebrevdo commented:
Jenkins, test this please.

danmane commented:
Thanks for the pull request!
